tests/bench/runtime_coreapis: add clist_sort() benchmark

maribu opened the pull request:
### Contribution description

This adds a benchmark for clist_sort() and calls that on three different list sizes.

### Testing procedure

Run the benchmark and check whether the run times are plausible.

### Issues/PRs references

None

crasbe commented:
Following the discussion from the Matrix chat, would it make sense to add a "reverse list" and "mostly sorted" list to the Benchmark as well?

maribu commented:
This is currently the reversed list (worst case) from the sorting algorithm's PoV. (All relevant have perfectly sorted in wrong order as worst case, or have no worst case).

The issue that Kaspar has is about cache locality: Right now two neighbouring nodes would also be adjacent in memory, as each cycle would now reverses the list, leaving topological neighbours also adjacent in mem.

For more realistical numbers, not having all list nodes comeing from the same buffer but being more scattered would be nices.

kaspar030 commented on tests/bench/runtime_coreapis/main.c:
```diff
@@ -58,6 +77,213 @@ static void _flag_waitone(void)
 
+static clist_node_t clist;
+
+static void _insert_reserved_data(clist_node_t *list)
```
May I suggest to abstract these into functions that can be used with `clist_foreach()`?

```
  static int _val_set_decrement(clist_node_t *node, void *arg)                                                                                                 
  {                                                                                                                                                            
      unsigned *val = arg;                                                                                                                                     
      struct test_node *tn = (struct test_node*)node;                                                                                                          
                                                                                                                                                               
      tn->value = *val;                                                                                                                                        
      *val -= 1;                                                                                                                                               
                                                                                                                                                               
      return 0;                                                                                                                                                
  }                                                                                                                                                            
                                                                                                                                                               
  static int _val_set_increment(clist_node_t *node, void *arg)                                                                                                 
  {                                                                                                                                                            
      unsigned *val = arg;                                                                                                                                     
      struct test_node *tn = (struct test_node*)node;                                                                                                          
                                                                                                                                                               
      tn->value = *val;                                                                                                                                        
      *val += 1;                                                                                                                                               
                                                                                                                                                               
      return 0;                                                                                                                                                
  }                                                                                                                                                            
                                                                                                                                                               
  static void _insert_reversed_data(clist_node_t *list)                                                                                                        
  {                                                                                                                                                            
      unsigned val = BENCH_CLIST_SORT_TEST_NODES;                                                                                                              
      clist_foreach(list, _val_set_decrement, &val);                                                                                                           
  }                                                                                                                                                            
                                                                                                                                                               
  static void _insert_sorted_data(clist_node_t *list)                                                                                                          
  {                                                                                                                                                            
      unsigned val = 0;                                                                                                                                        
      clist_foreach(list, _val_set_increment, &val);                                                                                                           
  }
```

This would have prevented the (now fixed) bug where `list->last` was skipped.
Also maybe with this in place, we can make the benchmark execution more parameterized:

```
static void _clist_sort_test(clist_node_t *list, prepare_data: fn(clist_node_t *list) {
/* (forgive me, forgot C function ptr syntax syntax) */
prepare_data(list);
clist_sort(list);
}
```

This would, maybe, down the line make it simpler to benchmark just the list preparation so  it doesn't skew the results.

kaspar030 commented:
Does the current merge sort behave very weird? As in, sorted data seems to be its worst case?

kaspar030 commented:
> The issue that Kaspar has is about cache locality: Right now two neighbouring nodes would also be adjacent in memory, as each cycle would now reverses the list, leaving topological neighbours also adjacent in mem.

Are you sure? I wasn't aware the nrf52840 has a memory cache.

I'm about to get pretty nerd sniped here. :)

kaspar030 commented on sys/include/benchmark.h:
```diff
@@ -44,6 +44,10 @@ extern "C" {
 #define BENCHMARK_FUNC(name, runs, func)                        \
     do {                                                        \
         ztimer_stopwatch_t timer = { .clock = ZTIMER_USEC };    \
+        /* warm up cache, branch predictor, ... */              \
```
Maybe keep this optional? Maybe:

```
#define BENCHMARK_FUNC_WARMUP(name, runs, warmup, func)
//...
#define BENCHMARK_FUNC(name, runs, func) BENCHMARK_FUNC_WARMUP(name, runs, 0, func)
```


Mostly because *sometimes* 32 runs might already be a lot (e.g., when testing `O(n^2)` algos on unrealistically large sets). And we don't want to make this adaptive here (like, specify warmup in milliseconds) I guess.

kaspar030 commented on tests/bench/runtime_coreapis/main.c:
```diff
@@ -79,69 +79,34 @@ static void _flag_waitone(void)
 static clist_node_t clist;
 
-static void _insert_reserved_data(clist_node_t *list)
+static int _insert_reserved_data(clist_node_t *item, void *arg)
```
```suggestion
static int _insert_reversed_data(clist_node_t *item, void *arg)
```

kaspar030 reviewed (approved):
This is now not using the warmup anymore. IMO it is not needed (did you observe values changing?). Anyhow, feel free to re-add that, split out the change to `benchmark.h`, or just leave it in. :)

The formatting is a bit off here:

```
2025-04-13 22:24:53,206 #      clist_sort (#4, rev):         7us  ---   7.000us per call  ---     142857 calls per sec    
2025-04-13 22:24:53,229 #     clist_sort (#4, prng):         7us  ---   7.000us per call  ---     142857 calls per sec    
2025-04-13 22:24:53,251 #     clist_sort (#4, sort):         7us  ---   7.000us per call  ---     142857 calls per sec    
2025-04-13 22:24:53,273 #  clist_sort (#4, ≈sort):         7us  ---   7.000us per call  ---     142857 calls per sec      
2025-04-13 22:24:53,296 #      clist_sort (#8, rev):        17us  ---  17.000us per call  ---      58823 calls per sec    
2025-04-13 22:24:53,318 #     clist_sort (#8, prng):        15us  ---  15.000us per call  ---      66666 calls per sec     
2025-04-13 22:24:53,340 #     clist_sort (#8, sort):        17us  ---  17.000us per call  ---      58823 calls per sec     
2025-04-13 22:24:53,362 #  clist_sort (#8, ≈sort):        16us  ---  16.000us per call  ---      62500 calls per sec       
```

... but I don't think that matters.

ACK.

kaspar030 commented:
Ah, I did a couple of runs with even larger sizes, with this and the insertion sort from #21398. The values make sense to me now. :+1:

maribu commented:
> > The issue that Kaspar has is about cache locality: Right now two neighbouring nodes would also be adjacent in memory, as each cycle would now reverses the list, leaving topological neighbours also adjacent in mem.
> 
> Are you sure? I wasn't aware the nrf52840 has a memory cache.

As far as I know the nRF52840 has no data cache, only an instruction cache (but I haven't looke d into the data sheet for a long time). I got confused with a test run on native.

maribu commented:
> The formatting is a bit off here:

I rephrased the `≈sort` to `alm.srt`. The approx sign is two bytes in length.

> This is now not using the warmup anymore. IMO it is not needed (did you observe values changing?). Anyhow, feel free to re-add that, split out the change to benchmark.h, or just leave it in. :)

I split it out. I think it helped a bit with the numbers on `native` being a bit more stable, but that might be luck.

mguetschow commented on tests/bench/runtime_coreapis/tests/01-run.py:
```diff
@@ -27,6 +27,11 @@ def testfunc(child):
+        child.expect(BENCHMARK_REGEXP.format(func=r"clist_sort \(#\d+, rev\)", timeout=TIMEOUT))
+        child.expect(BENCHMARK_REGEXP.format(func=r"clist_sort \(#\d+, prng\)", timeout=TIMEOUT))
+        child.expect(BENCHMARK_REGEXP.format(func=r"clist_sort \(#\d+, sort\)", timeout=TIMEOUT))
+        child.expect(BENCHMARK_REGEXP.format(func=r"clist_sort \(#\d+, ≈sort\)", timeout=TIMEOUT))
```
```suggestion
        child.expect(BENCHMARK_REGEXP.format(func=r"clist_sort \(#\d+, alm.srt\)", timeout=TIMEOUT))
```

?

mguetschow commented on tests/bench/runtime_coreapis/tests/01-run.py:
```diff
@@ -27,6 +27,11 @@ def testfunc(child):
     child.expect(BENCHMARK_REGEXP.format(func="thread flags set/wait one"), timeout=TIMEOUT)
     child.expect(BENCHMARK_REGEXP.format(func=r"msg_try_receive\(\)"), timeout=TIMEOUT)
     child.expect(BENCHMARK_REGEXP.format(func=r"msg_avail\(\)"))
+    for _ in range(7):
```
This depends on `BENCH_CLIST_SORT_TEST_NODES` I guess?

maribu replied:
Yes. I updated the test to print on `BENCH_CLIST_SORT_TEST_NODES` and the python script to expect the right values.

mguetschow commented on tests/bench/runtime_coreapis/main.c:
```diff
@@ -58,6 +77,170 @@ static void _flag_waitone(void)
+    struct test_node *n = container_of(item, struct test_node, list);
+    *seed = _prng_next(*seed);
+
+    /* By masking off some bits from the full sequence PRNG input we introduce
+     * duplicates: 2 dups in 64 items, 3 in 128 items, 20 in 256 items.
```
where do these numbers come from? (sorry, not that experienced with prng)

mguetschow commented on tests/bench/runtime_coreapis/main.c:
```diff
@@ -58,6 +77,170 @@ static void _flag_waitone(void)
+    /* By masking off some bits from the full sequence PRNG input we introduce
+     * duplicates: 2 dups in 64 items, 3 in 128 items, 20 in 256 items.
+     *
+     * The API explicitly says stable sort, so we better have some input data
+     * where this attribute would make a difference. */
```
How do we check the relative order of same nodes stays the same?

maribu replied:
Now we do. After the first sort when the list nodes are still in ascending memory order, we can check that nodes with the same value are still in ascending memory order with regard to each other.

After the benchmark we cannot assume this, as the insertion of data before each benchmark iteration just keeps the linked list's memory layout as it currently is.

I split out a dedicated test so that we can rely on the memory order for testing the sort to be stable.

mguetschow replied:
> I split out a dedicated test so that we can rely on the memory order for testing the sort to be stable.

As this part now tests for correctness, would it make more sense to have that in `tests/unittests` as in #21402 instead of `tests/bench`?

maribu replied:
It is also in the unit tests. So far, this test has not caught any error that the unit tests did not. But since we have different test data here anyway, I thought adding a few lines to also check for correctness might be a good idea.

mguetschow replied:
Okay, fine with me :)

mguetschow commented on tests/bench/runtime_coreapis/main.c:
```diff
@@ -58,6 +77,170 @@ static void _flag_waitone(void)
+
+static void _build_test_clist(size_t len)
+{
+
```
```suggestion
```

mguetschow commented on tests/bench/runtime_coreapis/main.c:
```diff
@@ -58,6 +77,170 @@ static void _flag_waitone(void)
+        struct test_node *lhs = container_of(prev, struct test_node, list);
+        struct test_node *rhs = container_of(cur, struct test_node, list);
+
+        if (lhs->value > rhs->value) {
```
You could use `_cmp` from above

maribu replied:
Good idea, that way one could change `_cmp()` and the check remains valid.

mguetschow reviewed (approved):
Thanks!

(Adding a benchmark test right before hard freeze shouldn't do any harm, should it? :yo_yo: )
